src/bluesearch/database/topic_info.py

```diff
@@ -0,0 +1,116 @@
+from typing import Any
+
+import bluesearch
+from bluesearch.database.article import ArticleSource
+
+
+@dataclass
+class TopicInfo:
+    """The topic information extracted from a journal article.
+
+    For the spec see the following GitHub issue/comment:
+    https://github.com/BlueBrain/Search/issues/518#issuecomment-985525160
+    """
+
+    source: ArticleSource
+    path: str | pathlib.Path
+    element_in_file: int | None = None
+    article_topics: dict[str, list[str]] = field(init=False, default_factory=dict)
+    journal_topics: dict[str, list[str]] = field(init=False, default_factory=dict)
+    metadata: dict = field(default_factory=dict)
+
+    def __post_init__(self) -> None:
+        """Run the post-initialization."""
+        self.creation_date = datetime.datetime.now()
+
+    @staticmethod
+    def add_topics(mapping: dict[str, list[str]], kind: str, topics: list[str]) -> None:
+        """Add topics to a mapping with collection of topics.
+
+        Parameters
+        ----------
+        mapping
+            A mapping of the form kind -> list-of-topics that shall be
+            updated in-place. For example ``{"MeSH": ["topic 1", "topic 2"]}``.
+        kind
+            The topic kind. Corresponds to a key in ``mapping``.
+        topics
+            The topics to add. Corresponds to a value in ``mapping``.
+        """
+        updated_topics = mapping.get(kind, []) + topics
+        mapping[kind] = sorted(set(updated_topics))
+
+    def add_article_topics(self, kind: str, topics: list[str]) -> None:
+        """Add article topics.
+
+        Parameters
+        ----------
+        kind
+            The topic kind. For example "MeSH" or "MAG".
+        topics
+            A list of the topics to add.
+        """
+        self.add_topics(self.article_topics, kind, topics)
+
+    def add_journal_topics(self, kind: str, topics: list[str]) -> None:
+        """Add journal topics.
+
+        Parameters
+        ----------
+        kind
+            The topic kind. For example "MeSH" or "MAG".
+        topics
+            A list of the topics to add.
+        """
+        self.add_topics(self.journal_topics, kind, topics)
+
+    def json(self) -> dict:
+        """Convert the contents of this class to a structured dictionary.
+
+        Returns
+        -------
+        dict
+            The structure dictionary with all topic information.
+        """
+        metadata: dict[str, Any] = {
+            **copy.deepcopy(self.metadata),
+            "created-date": self.creation_date.strftime("%Y-%m-%d %H:%M:%S"),
+            "bbs-version": bluesearch.__version__,
+        }
+        if self.element_in_file is not None:
+            metadata["element_in_file"] = self.element_in_file
+
+        json = {
+            "source": self.source.value,
+            "path": str(self.path),
+            "topics": {
+                "article": copy.deepcopy(self.article_topics),
+                "journal": copy.deepcopy(self.journal_topics),
```

Comment:
1. Just for me to understand, why do you think it is needed to do a `deepcopy` here ? 
2. Just to put into light, implementing the `topics` value this way means that we will have an `article` and `journal` field even when we do not search for one of them. Do you think it is a problem ?

Reply:
1. It's to make sure you can't mutate your `TopicInfo` by changing the returned dictionary. Without deepcopy if your do `json = topic_info.json(); json["topic"]["article"]["x"] = 5` then this will also change the internal state of `topic_info`.
2. Yes, this is actually a slightly different behaviour to what we used to do. Do you know if we rely on the presence/abcense of the "article"/"journal" keys? I couldn't see it specified anywhere. I reconed it would be easier to deal with if it's always present and of the same type, so you don't have to constantly check for Nones.

Reply:
1. I see, makes sense. Thanks!
2. I don't think we currently directly rely on this. I thought it was going to be though to distinguish when we did not look for a topic and when looked for it but did not find any. But in the first case, the dictionary is going to be empty and in the second one, the dictionary is going to have a key with the value being an empty list. So maybe it is not really a problem. Moreover, we can still check the bbs code version.